File: exemplos/sincronizacao/semaforos/python/threads_and_semaphores.py
from threading import *         
import time        
  
obj = Semaphore(1)        
  
#Using output
def show(name):    
    obj.acquire()                
    for i in range(5):
        print('Hello, ', name)
        time.sleep(1)
    obj.release()    
          
# creating six threads 
t1 = Thread(target = show , args = ('Thread-1',))
t2 = Thread(target = show , args = ('Thread-2',))
t3 = Thread(target = show , args = ('Thread-3',))
t4 = Thread(target = show , args = ('Thread-4',))
t5 = Thread(target = show , args = ('Thread-5',))
t6 = Thread(target = show , args = ('Thread-6',))
  
# calling the threads 
t1.start()
t2.start()
t3.start()
t4.start()
t5.start()
t6.start()

File: exemplos/sincronizacao/semaforos/C++/threads_and_semaphores.c
// C program to demonstrate working of Semaphores
#include <stdio.h>
#include <pthread.h>
#include <semaphore.h>
#include <unistd.h>
  
sem_t semaphore;
  
void* show(void* arg)
{
    char* name = (char*) arg;
    //wait
    sem_wait(&semaphore);
    for (int i =0; i<5; i++){
        printf("Hello %s\n", name);
        sleep(1);
    }
    sem_post(&semaphore);
}
  
  
int main()
{
    sem_init(&semaphore, 0, 2);
    pthread_t t1,t2,t3,t4,t5,t6;
    pthread_create(&t1,NULL,show,"Thread-1");
    pthread_create(&t2,NULL,show,"Thread-2");
    pthread_create(&t3,NULL,show,"Thread-3");
    pthread_create(&t4,NULL,show,"Thread-4");
    pthread_create(&t5,NULL,show,"Thread-5");
    pthread_create(&t6,NULL,show,"Thread-6");

    pthread_join(t1,NULL);
    pthread_join(t2,NULL);
    pthread_join(t3,NULL);
    pthread_join(t4,NULL);
    pthread_join(t5,NULL);
    pthread_join(t6,NULL);
    sem_destroy(&semaphore);
    return 0;
}